csrc/alias_analysis.cpp

```diff
@@ -395,52 +395,44 @@ void AliasAnalysisResult::add(
       i->second.first);
 }
 
-TensorView* AliasAnalysisResult::findNearestAliasedIo(
-    TensorView* fusion_out) const {
-  TensorView* root = fusion_out;
-  do {
-    const auto i = alias_to_source_.find(root);
-    root = (i == alias_to_source_.end() ? nullptr : i->second.first);
-  } while (root != nullptr && !root->isFusionInput() &&
-           !root->isFusionOutput());
-  return root;
-}
-
 TensorView* AliasAnalysisResult::getNearestAliasedIo(
-    const TensorView* fusion_out) const {
-  const auto i = out_to_root_.find(fusion_out);
-  return i == out_to_root_.end() ? nullptr : i->second;
+    const TensorView* alias) const {
+  const auto i = alias_to_root_.find(alias);
+  return i == alias_to_root_.end() ? nullptr : i->second;
 }
 
 namespace {
 bool okToRelayout(
-    const TensorView* out,
+    const TensorView* tv,
     const Layout& new_layout,
     const bool can_override_empty_allocation_domain) {
-  const std::vector<IterDomain*> out_allocation =
-      (can_override_empty_allocation_domain ? out->getAllocationDomain()
-                                            : out->getMaybeAllocationDomain());
-  return new_layout.isCompliantWith({out_allocation, out->getContiguity()});
+  const std::vector<IterDomain*> allocation =
+      (can_override_empty_allocation_domain ? tv->getAllocationDomain()
+                                            : tv->getMaybeAllocationDomain());
+  return new_layout.isCompliantWith({allocation, tv->getContiguity()});
 }
 } // namespace
 
 void AliasAnalysisResult::finalize(
-    Fusion* fusion,
     const bool can_override_empty_allocation_domain) {
-  for (TensorView* out :
-       ir_utils::filterByType<TensorView>(fusion->outputs())) {
-    TensorView* root = findNearestAliasedIo(out);
+  for (auto [alias, root_and_layout] : alias_to_source_) {
+    auto [root, preferred_layout] = root_and_layout;
+    // Walks up the `alias_to_source_` chain.
+    while (root != nullptr && !root->isFusionInput() &&
```

Comment:
nitpick: this walk up is maybe something we can simply cache. But this isn't that hot a path.